File: old/src/main/java/com/guider/healthring/wxapi/WXEntryActivity.java
package com.guider.healthring.wxapi;


import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import com.google.gson.Gson;
import com.guider.healthring.Commont;
import com.guider.healthring.MyApp;
import com.guider.healthring.activity.GuiderWxBindPhoneActivity;
import com.guider.healthring.bean.BlueUser;
import com.guider.healthring.bean.GuiderWXUserInfoBean;
import com.guider.healthring.bean.WXUserBean;
import com.guider.healthring.siswatch.NewSearchActivity;
import com.guider.healthring.siswatch.WatchBaseActivity;
import com.guider.healthring.siswatch.utils.WatchUtils;
import com.guider.healthring.util.Common;
import com.guider.healthring.util.OkHttpTool;
import com.guider.healthring.util.SharedPreferencesUtils;
import com.guider.healthring.util.ToastUtil;
import com.guider.healthring.util.URLs;
import com.guider.healthring.w30s.utils.httputils.RequestPressent;
import com.guider.healthring.w30s.utils.httputils.RequestView;
import com.tencent.mm.sdk.openapi.BaseReq;
import com.tencent.mm.sdk.openapi.BaseResp;
import com.tencent.mm.sdk.openapi.IWXAPI;
import com.tencent.mm.sdk.openapi.IWXAPIEventHandler;
import com.tencent.mm.sdk.openapi.SendAuth;
import com.tencent.mm.sdk.openapi.WXAPIFactory;
import org.json.JSONObject;
import java.util.HashMap;
import java.util.Map;


/**
 * MicroMsg
 * Created by wyl on 2017/2/6.
 */

public class WXEntryActivity extends WatchBaseActivity implements IWXAPIEventHandler,RequestView {

    private static final String TAG = "WXEntryActivity";

    //盖德微信登录
    private String GUIDER_WX_LOGIN_URL = "http://api.guiderhealth.com/api/v2/third/login/wachat/tokeninfo";

    private IWXAPI iwxapi;
    /**
     * 国内用这个
     */
    private static final String APP_SECRET = Commont.WX_APP_SECRET;
    public static final String WEIXIN_APP_ID = Commont.WX_APP_ID;

    /**
     * Google Play用这个
     */
//    private static final String APP_SECRET = "8864cc82ce546270207e5ba1f3ef0054";
//    public static final String WEIXIN_APP_ID = "wx7e5e9e90ae4d8f51";

    private RequestPressent requestPressent ;


    private WXUserBean wxUserBean = null;



    @Override
    public void onReq(BaseReq baseReq) {
//        Log.e(TAG, "---baseReq=" + baseReq.toString());
    }

    @Override
    public void onResp(BaseResp resp) {

        if(requestPressent == null)
            requestPressent = new RequestPressent();
            requestPressent.attach(this);

        if (resp instanceof SendAuth.Resp) {
            SendAuth.Resp newResp = (SendAuth.Resp) resp;

            Log.e(TAG,"---------wx="+newResp.userName+"\n"+newResp.token+"\n"+newResp.state);

            //获取微信传回的code  e9f2b83c
            String code = newResp.token;

            String wxResponseUrl = "https://api.weixin.qq.com/sns/oauth2/access_token?appid="
                    + WEIXIN_APP_ID + "&secret=" + APP_SECRET + "&code=" + code + "&grant_type=authorization_code";
            Log.e(TAG,"-------url="+wxResponseUrl);
            requestPressent.getRequestJSONObject(1001,wxResponseUrl,WXEntryActivity.this,1);


//            Log.e(TAG, "------code--" + code);

            /**
             * 通过这个获取信息
             * https://api.weixin.qq.com/sns/oauth2/access_token?appid=APPID&secret=SECRET&code=CODE&grant_type=authorization_code
             * 获得opid和access_token
             *0da2a2a2c3e35d25460c8f6bb01cd876
             */

        }
    }



    //注册到博之轮平台账号
    private void loginBZLForWx(WXUserBean wxUserBean) {
       // User wxUserBean = new Gson().fromJson(users, User.class);
        Log.e(TAG,"------wxUserBean="+wxUserBean.toString());
        Map<String,Object> params = new HashMap<>();
        params.put("thirdId", wxUserBean.getOpenid());
        params.put("thirdType",  "3");
        params.put("image", wxUserBean.getHeadimgurl());
        params.put("sex", (wxUserBean.getSex().equals("1")?"M":"F"));
        params.put("nickName", wxUserBean.getNickname());
        if (Commont.isDebug)Log.e(TAG, "3333游客注册或者登陆参数：" + params.toString());
        String url = URLs.HTTPs + URLs.disanfang;
        if (Commont.isDebug)Log.e(TAG, "====  json  " +  new Gson().toJson(params));
        OkHttpTool.getInstance().doRequest(url, new Gson().toJson(params), this, new OkHttpTool.HttpResult() {
                    @Override
                    public void onResult(String result) {
                        Log.e(TAG, "-------微信登录到bl=" + result);
                            try {
                                JSONObject jsonObject = new JSONObject(result);
                                if(jsonObject.has("resultCode")){
                                    if(jsonObject.getString("resultCode").equals("001")){
                                        String shuzhu = jsonObject.getString("userInfo");
                                        JSONObject jsonObjectStr = new JSONObject(shuzhu);
                                        String userId = jsonObjectStr.getString("userId");
                                        Gson gson = new Gson();
                                        BlueUser userInfo = gson.fromJson(shuzhu, BlueUser.class);
                                        Common.userInfo = userInfo;
                                        Common.customer_id = userId;
                                        //保存userid
                                        SharedPreferencesUtils.saveObject(WXEntryActivity.this, "userId", userInfo.getUserId());
                                    }

                                }
                            } catch (Exception E) {
                                E.printStackTrace();
                            }
                        }
                    });

    }


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
//        Log.e(TAG, "-----oncratewx---");
        //注册API
        iwxapi = WXAPIFactory.createWXAPI(this, WEIXIN_APP_ID, true);
        iwxapi.handleIntent(getIntent(), this);

        requestPressent = new RequestPressent();
        requestPressent.attach(this);

    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        setIntent(intent);
        iwxapi.handleIntent(intent, this);

    }


    @Override
    protected void onDestroy() {
        super.onDestroy();
        if(requestPressent != null)
            requestPressent.detach();
    }



    @Override
    public void showLoadDialog(int what) {

    }

    @Override
    public void successData(int what, Object object, int daystag) {
        if(object == null)
            return;
        Log.e(TAG,"--------Obj="+what+object.toString());
        switch (what){
            case 1001:  //获取微信的token
                analysiWXMsg(object.toString());
                break;
            case 1002:  //用户的信息
                analysisWXUserInfo(object.toString());
                break;
            case 1003:  //微信登录到盖德平台返回
                analysiWXToGuiderData(object.toString());
                break;
        }


    }



    @Override
    public void failedData(int what, Throwable e) {

    }

    @Override
    public void closeLoadDialog(int what) {

    }

    //解析微信的信息
    private void analysiWXMsg(String wxStr) {
        try {
            JSONObject jsonObject = new JSONObject(wxStr);
            if(jsonObject.has("access_token")){
                String wxAccessToken = jsonObject.getString("access_token");
                String wxOpenId = jsonObject.getString("openid");

                //获取用户的信息
                String userUrl = "https://api.weixin.qq.com/sns/userinfo?access_token=" + wxAccessToken + "&openid=" + wxOpenId;
                requestPressent.getRequestJSONObject(1002,userUrl,WXEntryActivity.this,2);

            }


        }catch (Exception e){
            e.printStackTrace();
        }
    }


    //解析用户信息，登录到盖德平台
    private void analysisWXUserInfo(String userStr) {
        wxUserBean = new Gson().fromJson(userStr, WXUserBean.class);
        WatchUtils.setIsWxLogin("LOGION_WX", new Gson().toJson(wxUserBean));

        loginBZLForWx(wxUserBean);
        Log.e(TAG,"------wxUsr="+wxUserBean.toString());
        Map<String,Object> maps = new HashMap<>();
        maps.put("appId","b3c327c04d8b0471");
        maps.put("headimgurl",wxUserBean.getHeadimgurl());
        maps.put("nickname",wxUserBean.getNickname());
        maps.put("openid",wxUserBean.getOpenid());
        maps.put("sex",Integer.valueOf(wxUserBean.getSex()));
        maps.put("unionid",wxUserBean.getUnionid());


        Log.e(TAG,"------ltMap="+new Gson().toJson(maps));
        requestPressent.getRequestJSONObject(1003,GUIDER_WX_LOGIN_URL,WXEntryActivity.this,new Gson().toJson(maps),3);

    }


    //解析登录到盖德返回信息，判断是否需要绑定手机号
    private void analysiWXToGuiderData(String str) {
        try {
            GuiderWXUserInfoBean guiderWXUserInfoBean = new Gson().fromJson(str,GuiderWXUserInfoBean.class);
            if(guiderWXUserInfoBean.getCode() != 0){
                ToastUtil.showToast(WXEntryActivity.this,guiderWXUserInfoBean.getMsg()+guiderWXUserInfoBean.getData());
                return;
            }
            boolean flag = guiderWXUserInfoBean.getData().isFlag();
            if(!flag){      //false不需要绑定手机号
                long accountId = guiderWXUserInfoBean.getData().getTokenInfo().getAccountId();
                SharedPreferencesUtils.setParam(MyApp.getInstance(), "accountIdGD", accountId);
                startActivity(NewSearchActivity.class);
                finish();
            }else{

                if(wxUserBean == null)
                    return;

                Map<String,Object> param = new HashMap<>();
                param.put("openId",wxUserBean.getOpenid());
                param.put("unionId",wxUserBean.getUnionid());
                param.put("headUrl",wxUserBean.getHeadimgurl());
                param.put("sex",wxUserBean.getSex());
                param.put("nickname",wxUserBean.getNickname());
                String wxStr = new Gson().toJson(param);

                startActivity(GuiderWxBindPhoneActivity.class,new String[]{"wxStr"},new String[]{wxStr});

            }


        }catch (Exception e){
            e.printStackTrace();
        }
    }

}

File: ring/src/main/java/com/guider/gps/wxapi/WXEntryActivity.kt
package com.guider.gps.wxapi

import android.content.Intent
import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import com.google.gson.Gson
import com.guider.baselib.base.BaseApplication
import com.guider.baselib.utils.*
import com.guider.feifeia3.utils.ToastUtil
import com.joinutech.ddbeslibrary.bean.WXAccessTokenEntity
import com.joinutech.ddbeslibrary.bean.WXBaseRespEntity
import com.joinutech.ddbeslibrary.bean.WXUserInfo
import com.tencent.mm.opensdk.modelbase.BaseReq
import com.tencent.mm.opensdk.modelbase.BaseResp
import com.tencent.mm.opensdk.openapi.IWXAPI
import com.tencent.mm.opensdk.openapi.IWXAPIEventHandler
import com.tencent.mm.opensdk.openapi.WXAPIFactory
import com.zhy.http.okhttp.OkHttpUtils
import com.zhy.http.okhttp.callback.StringCallback
import okhttp3.Call

class WXEntryActivity : AppCompatActivity(), IWXAPIEventHandler {
    private lateinit var api: IWXAPI
    private val RETURN_MSG_TYPE_LOGIN = 1
    private val RETURN_MSG_TYPE_SHARE = 2

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        api = WXAPIFactory.createWXAPI(this, APP_ID_WX, true)
        api.registerApp(APP_ID_WX)
        //注意：
        //第三方开发者如果使用透明界面来实现WXEntryActivity，需要判断handleIntent的返回值，
        // 如果返回值为false，则说明入参不合法未被SDK处理，应finish当前透明界面，
        // 避免外部通过传递非法参数的Intent导致停留在透明界面，引起用户的疑惑
        val result = api.handleIntent(intent, this)
        if (!result) {
            logShow("参数不合法，未被SDK处理，退出")
            finish()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        api.handleIntent(data, this)
    }

    override fun onNewIntent(intent: Intent?) {
        super.onNewIntent(intent)
        setIntent(intent)
        api.handleIntent(intent, this)
        finish()
    }

    private fun logShow(text: String) {
        Log.d("WXLogin", text)
    }

    override fun onReq(baseReq: BaseReq?) {
        logShow("baseReq: ${Gson().toJson(baseReq)}")
    }

    override fun onResp(baseReq: BaseResp?) {
        val json = Gson().toJson(baseReq)
        logShow("baseReq: ${Gson().toJson(baseReq)}")
        logShow("baseResp--B:${baseReq!!.errStr},${baseReq.openId}," +
                "${baseReq.transaction},${baseReq.errCode}")
        val entity: WXBaseRespEntity = Gson().fromJson(json, WXBaseRespEntity::class.java)
        val result: String
        //根据errCode判断情况
        when (baseReq.errCode) {
            BaseResp.ErrCode.ERR_OK -> {
                //用户同意
                result = "发送成功"
                OkHttpUtils.get().url("https://api.weixin.qq.com/sns/oauth2/access_token")
                        .addParams("appid", APP_ID_WX)
                        .addParams("secret", APP_SECRET_WX)
                        .addParams("code", entity.code)
                        .addParams("grant_type", "authorization_code")
                        .build()
                        .execute(object : StringCallback() {
                            override fun onResponse(response: String?, id: Int) {
                                logShow("response:$response")
                                val accessTokenEntity: WXAccessTokenEntity? = Gson()
                                        .fromJson(response, WXAccessTokenEntity::class.java)
                                if (accessTokenEntity != null) {
                                    val openid = accessTokenEntity.openid
                                    val unionid = accessTokenEntity.unionid
                                    logShow("微信登录资料已获取，后续未完成")
                                    if (StringUtil.isNotBlankAndEmpty(MMKVUtil.getString("wxInfo"))) {
                                        //需要用户信息
                                        getUserInfo(accessTokenEntity)
                                    } else {
                                        //只是登录
                                        EventBusUtils.sendEvent(EventBusEvent(
                                                EventBusAction.WECHAT_LOGIN,
                                                "$openid:$unionid"))
                                        finish()
                                    }
                                } else {
                                    logShow("获取失败")
                                }
                            }

                            override fun onError(p0: Call?, p1: Exception?, p2: Int) {
                                logShow("请求错误")
                            }

                        })
            }
            BaseResp.ErrCode.ERR_USER_CANCEL -> {
                //用户取消
                result = "发送取消"
                if (baseReq.type == RETURN_MSG_TYPE_SHARE) {
                    ToastUtil.show(BaseApplication.guiderHealthContext, "微信分享取消")
                } else {
                    ToastUtil.show(BaseApplication.guiderHealthContext, "微信登录取消")
                }
                finish()
            }
            BaseResp.ErrCode.ERR_AUTH_DENIED -> {
                //用户拒绝授权
                result = "发送被拒绝"
                ToastUtil.show(BaseApplication.guiderHealthContext, "微信登录失败")
                finish()
            }
            BaseResp.ErrCode.ERR_BAN -> {
                //签名错误
                result = "签名错误"
                ToastUtil.show(BaseApplication.guiderHealthContext, "微信登录失败")
                finish()
            }
            else -> {
                result = "发送返回"
                ToastUtil.show(BaseApplication.guiderHealthContext, "微信登录取消")
                finish()
            }
        }
        logShow(result)
    }

    private fun getUserInfo(accessTokenEntity: WXAccessTokenEntity) {
        OkHttpUtils.get()
                .url("https://api.weixin.qq.com/sns/userinfo")
                .addParams("access_token", accessTokenEntity.access_token)
                .addParams("openid", accessTokenEntity.openid)//openid:授权用户唯一标识
                .build()
                .execute(object : StringCallback() {
                    override fun onResponse(response: String?, id: Int) {
                        logShow("response:$response")
                        val wxResponse: WXUserInfo? = Gson().fromJson(response,
                                WXUserInfo::class.java)
                        var nickName = ""
                        var headimgurl = ""
                        if (wxResponse != null) {
                            if (StringUtil.isNotBlankAndEmpty(wxResponse.nickname)) {
                                nickName = wxResponse.nickname!!
                            }
                            if (StringUtil.isNotBlankAndEmpty(wxResponse.nickname)) {
                                headimgurl = wxResponse.headimgurl!!
                            }
                            EventBusUtils.sendEvent(EventBusEvent(
                                    EventBusAction.WECHAT_LOGIN,
                                    "${wxResponse.openid}:" +
                                            "${wxResponse.unionid}:" +
                                            "${wxResponse.sex}:" +
                                            "$nickName:" +
                                            headimgurl))
                        }
                        finish()
                    }

                    override fun onError(p0: Call?, p1: Exception?, p2: Int) {
                        logShow("请求错误")
                    }

                })
    }

}